sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuShuffleCoalesceExec.scala

```diff
@@ -194,6 +197,70 @@ class JCudfTableOperator extends SerializedTableOperator[SerializedTableColumn]
   }
 }
 
+case class KudoHostMergeResultWrapper(inner: Either[KudoHostMergeResult, Int],
+    dataSize: Long) extends CoalescedHostResult {
```

Comment:
I find Either to be tricky to read in practice, and in this case it's cleaner to let old fashioned object-oriented programming handle the differences rather than if checks in the methods of a single class.  The caller already has to do a check to figure out which type of Either to build, so have the caller build different classes (which also removes the need to pass dummy values in the row-only case), e.g.:
```scala
case class KudoHostMergeResultWrapper(inner: KudoHostMergeResult) extends CoalescedHostResult {
  override def toGpuBatch(dataTypes: Array[DataType]): ColumnarBatch = {
    RmmRapidsRetryIterator.withRetryNoSplit {
    closeOnExcept(result.toTable) { cudfTable =>
      GpuColumnVector.from(cudfTable, dataTypes)
    }
  }

  override def getDataSize: Long = inner.getDataLength

  override def close(): Unit = inner.close()
}

case class RowCountOnlyMergeResult(rowCount: Int) extends CoalescedHostResult {
  override def toGpuBatch(dataTypes: Array[DataType]): ColumnarBatch = {
    new ColumnarBatch(Array.empty, rowCount)
  }

  override def getDataSize: Long = 0

  override def close(): Unit = {}
}

[...]

  override def concatOnHost(columns: Array[KudoSerializedTableColumn]): CoalescedHostResult = {
    require(columns.nonEmpty, "no tables to be concatenated")
    val numCols = columns.head.kudoTable.getHeader.getNumColumns
    if (numCols == 0) {
      val totalRowsNum = columns.map(getNumRows).sum
      RowCountOnlyMergeResult(totalRowsNum)
    } else {
      val kudoTables = new util.ArrayList[KudoTable](columns.length)
      columns.foreach { column =>
        kudoTables.add(column.kudoTable)
      }

      val result = kudo.mergeOnHost(kudoTables)
      kudoMergeHeaderTime += result.getRight.getCalcHeaderTime
      kudoMergeBufferTime += result.getRight.getMergeIntoHostBufferTime

      KudoHostMergeResultWrapper(result.getLeft)
    }
  }
```